### ISSUE TITLE
Backslash on sign  multiplication

### ISSUE THREAD

**Issue by falkse** [cuberite/cuberite#1965]
By finishing editing a sign, one backslash will be twice on the sign. And so on with 2 backslashes, there will be 4.
Count of backslash \* count of backlash = count of backslashes on the sign.

**Comment by bearbin** [comment]
Would it be possible to provide a screenshot (or video) of the behaviour that you're experiencing?

**Comment by falkseidl** [comment]
Yes of course:
https://www.youtube.com/watch?v=_uee5Mr-7yk&feature=youtu.be

**Comment by bearbin** [comment]
Thanks! I'll look into what's actually causing the issue.

**Comment by falkseidl** [comment]
Thank you!

**Comment by rcrimp** [comment]
Bug is easily reproducable. 
I'm on Fedora 21, compiled from git master locally, running server and client from the same machine.

I suspect that the backslash character '\' is being escaped to '\\', but it's not being un-escaped. I'm going to look into this a bit further, if that's alright with you @bearbin.

**Comment by rcrimp** [comment]
Yep, characters aren't being un-escaped properly.
http://imgur.com/a/244Qu
*&* becomes *\u0026* and *=* becomes *\u003d*

and I'm sure there are many other afflicted characters.

**Comment by rcrimp** [comment]
The issue was introduced here with the JSON writer https://github.com/mc-server/MCServer/commit/12621e3800ec353bca179243d521580e0da6eae9

Abandoning the JSON writer and reverting back to hardcoding the json syntax fixes the issue
````C++
AString Lines[] = { a_Line1, a_Line2, a_Line3, a_Line4 };
for (Astring& line : Lines)
{
    Pkt.WriteString(Printf("{\"text\": \"%s\"}", line.c_str()));
}
````

If this is an adequate fix, then I can issue a pull request.
If not could you point me in the right direction for a better fix, this is my first attempt at helping with an open source project.

**Comment by madmaxoft** [comment]
The problem with reverting back is that the sign text can come from various sources, not only the client setting the sign text, but also a plugin etc., and those won't do the json escaping.

The proper fix would be to have a look why the strings aren't unescaped when receiving the SetSignText packet from the client.

**Comment by rcrimp** [comment]
Yep, you're right, the strings aren't being unescaped on the server end.

'*\\u0026*' is received and stored, which is then escaped to '*\\\\u0026*' before being sent back to the client, and that double backslash is the issue.

cProtocol180::HandlePaketUpdateSign should probably unescape the string *Line*. We can use jsoncpp like this (but it's a bit awkward). 

````git
+       Json::Reader reader;
        AString Lines[4];
        for (int i = 0; i < 4; i++)
        {
                HANDLE_READ(a_ByteBuffer, ReadVarUTF8String, AString, Line);
-               Lines[i] = Line.substr(1, Line.length() - 2);  // Remove ""
+               AString temp1 = Line.substr(1, Line.length() - 2);  // Remove ""

+               Json::Value root;
+               reader.parse(Printf("{\"text\": \"%s\"}", temp1.c_str()), root, false);
+               Lines[i] = root.get("text", "Unknown").asString();
        }
````
This fixes the issue but a more elegant (and likely faster) solution would be to add a decodeString method to the StringUtils and unescape the characters ourselves. jsoncpp has a very similar method which could easily be adapted to work for std::Strings.

**Comment by peterbell10** [comment]
Fixed by 1fb89a540950e76f8f0f5ec08eceb5e41197910a.
